crates/client/mempool/src/inner.rs

```diff
@@ -0,0 +1,536 @@
+                        let removed = self
+                            .tx_queue
+                            .remove(&AccountOrderedByTimestamp { contract_addr, timestamp: former_head_arrived_at });
+                        debug_assert!(removed);
+                        let inserted =
+                            self.tx_queue.insert(AccountOrderedByTimestamp { contract_addr, timestamp: arrived_at });
+                        debug_assert!(inserted);
+                    }
+                    InsertedPosition::Other => {
+                        // No need to update the tx queue.
+                    }
+                }
+            }
+            hash_map::Entry::Vacant(entry) => {
+                // Insert the new nonce chain
+                let nonce_chain = NonceChain::new_with_first_tx(mempool_tx);
+                entry.insert(nonce_chain);
+
+                // Also update the tx queue.
+                let inserted = self.tx_queue.insert(AccountOrderedByTimestamp { contract_addr, timestamp: arrived_at });
+                debug_assert!(inserted);
+            }
+        };
+        Ok(())
+    }
+
+    pub fn has_deployed_contract(&self, addr: &ContractAddress) -> bool {
+        self.deployed_contracts.contains(addr)
+    }
+
+    pub fn pop_next(&mut self) -> Option<MempoolTransaction> {
+        // Pop tx queue.
+        let tx_queue_account = self.tx_queue.pop_first()?; // Bubble up None if the mempool is empty.
+
+        // Update nonce chain.
+        let nonce_chain =
+            self.nonce_chains.get_mut(&tx_queue_account.contract_addr).expect("Nonce chain does not match tx queue");
+        let (mempool_tx, nonce_chain_new_state) = nonce_chain.pop();
+        match nonce_chain_new_state {
+            NonceChainNewState::Empty => {
+                // Remove the nonce chain.
+                let removed = self.nonce_chains.remove(&tx_queue_account.contract_addr);
+                debug_assert!(removed.is_some());
+            }
+            NonceChainNewState::NotEmpty => {
+                // Re-add to tx queue.
+                let inserted = self.tx_queue.insert(AccountOrderedByTimestamp {
+                    contract_addr: tx_queue_account.contract_addr,
+                    timestamp: nonce_chain.front_arrived_at,
+                });
+                debug_assert!(inserted);
+            }
+        }
+
+        // Update deployed contracts.
+        if let AccountTransaction::DeployAccount(tx) = &mempool_tx.tx {
+            let removed = self.deployed_contracts.remove(&tx.contract_address);
+            debug_assert!(removed);
+        }
+
+        Some(mempool_tx)
+    }
+
+    pub fn pop_next_chunk(&mut self, dest: &mut Vec<MempoolTransaction>, n: usize) {
+        for _ in 0..n {
+            let Some(tx) = self.pop_next() else { break };
+            dest.push(tx);
+        }
+    }
+
+    pub fn readd_txs(&mut self, txs: Vec<MempoolTransaction>) {
```

Comment:
ah, it's re_add not read_d
the block production pops transactions as chunks, so that we can support optimistic parallelization
but there is the case where the block gets full before the end of the chunk, and in this case you need to re-add transactions to the mempool, with a special `force` mode to handle nonce collision.

i'll change the function name and add a comment about that